nebula_hw_interfaces/include/nebula_hw_interfaces/nebula_hw_interfaces_common/connections/udp.hpp

```diff
@@ -0,0 +1,383 @@
+    if (bytes > static_cast<size_t>(INT32_MAX))
+      throw UsageError("The maximum value supported (0x7FFFFFF) has been exceeded");
+
+    auto buf_size = static_cast<int>(bytes);
+    int result = setsockopt(sock_fd_, SOL_SOCKET, SO_RCVBUF, &buf_size, sizeof(buf_size));
+    if (result < 0) throw SocketError(errno);
+    return *this;
+  }
+
+  /**
+   * @brief Join an IP multicast group. Only one group can be joined by the socket.
+   *
+   * @param group_ip The multicast IP. It has to be in the multicast range `224.0.0.0/4` (between
+   * `224.0.0.0` and `239.255.255.255`).
+   */
+  UdpSocket & join_multicast_group(const std::string & group_ip)
+  {
+    if (state_ < State::INITIALIZED) throw UsageError("Socket has to be initialized first");
+    if (state_ >= State::BOUND)
+      throw UsageError("Multicast groups have to be joined before binding");
+
+    ip_mreq mreq;
+    mreq.imr_multiaddr.s_addr = inet_addr(group_ip.c_str());  // Multicast group address
+    mreq.imr_interface.s_addr = inet_addr(host_.ip.c_str());
+
+    int result = setsockopt(sock_fd_, IPPROTO_IP, IP_ADD_MEMBERSHIP, &mreq, sizeof(mreq));
+    if (result < 0) throw SocketError(errno);
+
+    multicast_ip_.emplace(group_ip);
+    return *this;
+  }
+
+  /**
+   * @brief Bind the socket to host IP and port given in `init()`. If `join_multicast_group()` was
+   * called before this function, the socket will be bound to `group_ip` instead. At least `init()`
+   * has to have been called before.
+   */
+  UdpSocket & bind()
+  {
+    if (state_ < State::INITIALIZED) throw UsageError("Socket has to be initialized first");
+    if (state_ >= State::BOUND) throw UsageError("Re-binding already bound socket");
+
+    sockaddr_in addr{};
+    addr.sin_family = AF_INET;
+    addr.sin_port = htons(host_.port);
+    addr.sin_addr.s_addr = multicast_ip_ ? inet_addr(multicast_ip_->c_str()) : INADDR_ANY;
+
+    int result = ::bind(sock_fd_, (struct sockaddr *)&addr, sizeof(addr));
+    if (result == -1) throw SocketError(errno);
+
+    state_ = State::BOUND;
+    return *this;
+  }
+
+  /**
+   * @brief Register a callback for processing received packets and start the receiver thread. The
+   * callback will be called for each received packet, and will be executed in the receive thread.
+   * Has to be called on a bound socket (`bind()` has to have been called before).
+   *
+   * @param callback The function to be executed for each received packet.
+   */
+  UdpSocket & subscribe(callback_t && callback)
+  {
+    if (state_ < State::BOUND) throw UsageError("Socket has to be bound first");
+    if (state_ > State::BOUND) throw UsageError("Cannot re-subscribe to socket");
+
+    callback_ = std::move(callback);
+    launch_receiver();
+    return *this;
+  }
+
+  /**
+   * @brief Gracefully stops the active receiver thread (if any) but keeps the socket alive. The
+   * same socket can later be subscribed again.
+   */
+  UdpSocket & unsubscribe()
```

Comment:
If the socket is kept alive, should we consider shutting it down to prevent reading from it?
`shutdown(sock_fd_, SHUT_RD);`
and associated error handling.

Reply:
Given that
1. the only reads made are in the now-stopped `receive_thread_`
2. the user might want to re-subscribe with the same socket later (kind of an esoteric argument maybe...)

I think it would best be left open. As far as I understand, since the socket is not actually `close`d by `shutdown`, there should not be any performance/resource difference between doing/not doing it here.